File: nodejs-typescript/src/domain/Reconciliation.ts
import type { CompanyAccountBook } from './CompanyAccountBook.js';
import { ReconciliationError } from './errors.js';

export function verifyReconciliation(accountBook: CompanyAccountBook): void {
  for (const snapshot of accountBook.finalBalances()) {
    if (snapshot.balance.isNegative()) {
      throw new ReconciliationError(
        `Negative balance on account ${snapshot.accountNumber.toString()}`,
      );
    }
  }
}

File: dotnet/src/MableBank.Domain/Reconciliation.cs
using MableBank.Domain.Errors;

namespace MableBank.Domain;

public static class Reconciliation
{
    public static void Verify(CompanyAccountBook accountBook)
    {
        ArgumentNullException.ThrowIfNull(accountBook);
        VerifyBalances(accountBook.FinalBalances());
    }

    public static void VerifyBalances(IEnumerable<AccountBalance> balances)
    {
        ArgumentNullException.ThrowIfNull(balances);
        foreach (var snapshot in balances)
        {
            if (snapshot.Balance.IsNegative())
            {
                throw new ReconciliationException(
                    $"Negative balance on account {snapshot.AccountNumber}");
            }
        }
    }
}
